### ISSUE TITLE
Calculate recovery period with linelist to complement recovery data in China during initial phase (Jan-Mar)

### ISSUE THREAD

**Issue by Inglezos** [lisphilar/covid19-sir#357]
## Summary of question
I wonder if we can make substantial analysis for the initial major pandemic wave in China and its provinces during February-May. I mean to analyze the data and fit then in SIRF model. Is that possible, considering that the initial phase is problematic during analysis and has big error? Could you share the code for such provinces analysis for that period?

**Comment by lisphilar** [comment]
Sorry, could you please explain your question in a different way because I could not understand the question? Please share the codes and outputs you got the question with.

**Comment by Inglezos** [comment]
- First of all, `subset()` fails to load correct recovered data for China. Please run this country analysis to double-check this with me.
- Apart from that, let's say I want to run China and some provenance analysis for the period January - March 2020, when the pandemic reached its peak in China. But the first phase is probably until late February so the time window of the data of that pandemic wave is too narrow to run substantial analysis. Is this correct?

**Comment by lisphilar** [comment]
> First of all, subset() fails to load correct recovered data for China. Please run this country analysis to double-check this with me.

`JHUData.subset("China")` raises ValueError, but `JHUData.subset_complement("China")` returns records. This is because recovered data is not registered in the raw dataset as well as Netherlands #334 .
https://gist.github.com/lisphilar/e5e61a8a9b5acab62bc635d46d200323

> But the first phase is probably until late February so the time window of the data of that pandemic wave is too narrow to run substantial analysis. Is this correct?

What is the problem? Why do you think? Please share the codes or figures to discuss.
- Do you mean the first phase in China is too long?
- Do you mean it is difficult to estimate and discuss parameter values because outbreak did not show complete end?

Note for the first one:
Long phase does not mean useless analysis.
In the initial phase, parameter values could be stable because we did not know how to prevent outbreak. Long phase only means that parameter values were not changed.

Note for the second one:
To cover the weak point of conventional SIR model analysis with fixed parameter values for long period, I introduced "phases". Phase-dependent SIR analysis allows change of parameter values.

**Comment by Inglezos** [comment]
> This is because recovered data is not registered in the raw dataset as well as Netherlands
>
In `jhu_data.py -> subset()`, at line `df = super().subset(country=country, province=province, start_date=start_date, end_date=end_date)`, the function fails to load the existing recovered data. I think that such data already exist normally and are not missing. I checked manually the `covid19dh.csv` file and I see that the first recovered data records for China begin on 20Jan20:
![image](https://user-images.githubusercontent.com/36845191/100261120-f63dc880-2f52-11eb-961f-66991493b4ff.png)

**Comment by lisphilar** [comment]
> The problem for China is that the analysis/simulation is made after 23Febr20, because the initial phase is skipped. The thing is that the first month was very critical for the China pandemic and if I want to fit to the model the first month (10Jan-28Febr) I cannot because of that. Is this correct?

It may be necessary to revise algorithm of complement.The first recovery date in China is estimated as 23Feb2020 with the following causes.

- The first confirmation date of China in the dataset is 04Jan2020,
- the calculated mode values of closing period (time to get outcome) is about 27 days, and
- the number of confirmed cases incleased sharply in Jan2020.

We can
- adjust value of closing period to fit the situation and/or
- use moving average values of confirmed cases when complement recovered data.

Do you have time to investigate `JHUData._complement_recovered_full()` method?


> Can I run full analysis for these provinces for example? i) For Jan-Febr and ii) for after February? I think provinces are supported right?

Yes, provinces are supported, like `cs.JHUData.subset(country="China", province="Hubei")` and`cs.Scenario(jhu_data, population_data, country="China", province="Hubei")`. Country name is required, but province name is optional in the methods.

https://gist.github.com/lisphilar/f1ec31e4a5078846612a4f6ce29ee1d5
(WIth Hubei data, estimation failed in the last phases because the number of infected cases is small. This should be fixed with a new issue.)


> Also I notice that for the initial phase in China, the Rt = 0.22.....Does this mean that the model is very good or that we have something like over-fitting of the data to the model?

Root cause would be the algorithm of complement.

**Comment by Inglezos** [comment]
I checked the `_complement_recovered_full()` method as well as `calculate_closing_period()`.
The only improvement I can think of is to use inside of `calculate_closing_period(), the 'std()` or `quantile(0.25)` instead of mode value for closing day and a rolling mean of 3 days for Confirmed cases:
```
    def calculate_closing_period(self):
        """
        Calculate mode value of closing period, time from confirmation to get outcome.

        Returns:
            int: closing period [days]
        """
        # Get cleaned dataset at country level
        df = self._cleaned_df.copy()
        df = df.loc[df[self.PROVINCE] == self.UNKNOWN]
        # Select records of countries where recovered values are reported
        df = df.groupby(self.COUNTRY).filter(lambda x: x[self.R].sum() != 0)
        # Total number of confirmed/closed cases of selected records
        df = df.groupby(self.DATE).sum()
        df[self.FR] = df[[self.F, self.R]].sum(axis=1)
        df = df.loc[:, [self.C, self.FR]]
        **df[self.C] = df[self.C].rolling(window=3).mean()**
        # Calculate how many days to confirmed, closed
        df = df.unstack().reset_index()
        df.columns = ["Variable", self.DATE, "Number"]
        df["Days"] = (df[self.DATE] - df[self.DATE].min()).dt.days
        df = df.pivot_table(values="Days", index="Number", columns="Variable")
        df = df.interpolate(limit_direction="both").fillna(method="ffill")
        df["Elapsed"] = df[self.FR] - df[self.C]
        df = df.loc[df["Elapsed"] > 0]
        # Calculate mode value of closing period
        **return df["Elapsed"].std().astype(np.int64)**
```

**Comment by Inglezos** [comment]
[UPDATE:]
I tried something else (without any changes mentioned in my previous comment).
In `_complement_recovered_full()`, the current code rejects all the negative calculated recovered cases. What if we utilized them with their absolute value and sorted? I modified the code to the following:
```
        # Closing period
        self._closing_period = self._closing_period or self.calculate_closing_period()
        # Estimate recovered records
        shifted = df[self.C].shift(periods=self._closing_period, freq="D")
        df[self.R] = shifted - df[self.F]
        df.loc[df[self.R] < 0, self.R] *= -1
        df[self.R].interpolate(method="spline", order=1, inplace=True)
        df[self.R] = sorted(df[self.R])
        df[self.R] = df[self.R].fillna(0).astype(np.int64)
```
The result from this, for China, is the start date to be 28Jan20, which is very good and the simulation has also good results. The only problem is the spikes and the non-monotonic increase of simulated fatal and recovered cases. This is another problem that we already have discussed in another issue. But what if we simply sort the simulated cases? Then it would be resolved.
I mean by putting in `scenario.simulate()` the following code:
```
        tracker = self._create_tracker(name)
        try:
            sim_df = tracker.simulate(y0_dict=y0_dict)
            sim_df[self.F] = sorted(sim_df[self.F])
            sim_df[self.R] = sorted(sim_df[self.R])
```

**Comment by lisphilar** [comment]
Thank you for your detailed investigation and simulation with new argument `sorted=True` will improved the simulated number of cases greatly.

However, sorting in `Scenario.simulate()` may mask the root cause, in-accurate parameter estimation in some phases, without solving it. Sorting will replace simulated number of cases over phases. RMSLE scores fo `Scenario.score()` and figures of `Scenario.simulate()` and `Scneario.history("Infected")` etc. will be greatly improved, but parameter values are still in-accurate in some phases. The number of cases is only observal variable and parameter values have physical meanings in SIR-derived model.

On the other hand, we can use sorting tools in `JHUData._complement_recovered_full()` because this is free from ODE models. The purpose of this method is to reflect real situations on the datasets.
`df.loc[df[self.R] < 0, self.R] *= -1` (or `df[self.R] = df[self.R].abs()`) and `df[self.R] = sorted(df[self.R])` (or `df[self.R].sort_values(inplace=True)`) will be effective solutions.

> Those linelist data when will be utilized? I thought that we would use them for generating the missing recovered cases.

We can use linelist data as follows. #281 
```Python
linelist = data_loader.linelist()
all_df = linelist.cleaned()
china_df = linelist.subset("China")
```

**Comment by Inglezos** [comment]
Yes I agree that sorting in `Scenario.simulate()` might camouflage the root cause; we need to eliminate the underlying problem.
I will make soon a pull request for sorting the recovered cases (in full complement as well as in `subset_complement`).

Regarding the linelist, I thought that there was already a method that was using these linelist data in order to extract approximated recovered data, similar to the complementary method but more accurate.

**Comment by lisphilar** [comment]
Thank you.

#360 and #362 will be closed to use linelist data like this.
```Python
linelist = data_loader.linelist()
df = linelist.closed("China", outcome="Recovered")
```

**Comment by Inglezos** [comment]
So in that case we won't use the complementary recovered data anymore? We will approximate the missing ones from such linelist data?

**Comment by lisphilar** [comment]
We could use linelist data, but only 36 records were found with `linelist.closed("China", outcome="Recovered")`. It will be difficult to use it for complement.


Confirmation_date | Recovered_date
-- | --
2020/1/21 | 2020/1/24
2020/1/21 | 2020/1/29
2020/1/24 | 2020/2/3
2020/1/25 | 2020/2/12
2020/1/26 | 2020/2/5
2020/1/26 | 2020/2/5
2020/1/26 | 2020/2/5
2020/1/27 | 2020/2/7
2020/1/31 | 2020/2/10
2020/2/2 | 2020/2/14
2020/2/5 | 2020/2/13
2020/2/4 | 2020/2/13
2020/2/4 | 2020/2/18
2020/2/8 | 2020/2/17
2020/2/3 | 2020/2/22
2020/1/25 | 2020/2/2
2020/1/24 | 2020/2/2
2020/1/24 | 2020/1/28
2020/2/3 | 2020/2/22
2020/1/25 | 2020/2/2
2020/1/24 | 2020/2/2
2020/1/26 | 2020/2/5
2020/1/26 | 2020/2/5
2020/1/26 | 2020/2/5
2020/1/24 | 2020/1/28
2020/1/21 | 2020/1/24
2020/1/27 | 2020/2/7
2020/1/24 | 2020/2/3
2020/1/31 | 2020/2/10
2020/2/2 | 2020/2/14
2020/1/21 | 2020/1/29
2020/2/5 | 2020/2/13
2020/2/4 | 2020/2/13
2020/2/4 | 2020/2/18
2020/1/25 | 2020/2/12
2020/2/8 | 2020/2/17

**Comment by Inglezos** [comment]
Yes I think the modus is around 10 days.
I edited my previous comment, the complementary method shall be kept, but the closing period will be affected from the linelist data?

**Comment by lisphilar** [comment]
Because the number of records is small, it is not recommended to use the linelist data for complement.

With my notebook created on , I found that the distribution of closing period was bimodal (10 days and 30 days).
https://gist.github.com/lisphilar/9c5f0a96c88a27be0c3e085ced89a2b1

![index](https://user-images.githubusercontent.com/7270139/100521188-66fc1380-31e5-11eb-914a-7aed0fec8571.png)

I do not have time in some days, but it will be good if we have reasons we use 10 days as representative value of closing period. I wonder whether closing period is different that in the first months of outbreak and that in these days or not.

**Comment by Inglezos** [comment]
Hmmm could it be one for the fatal closing date and one period for the actual recovered date? I mean, during closing period calculation, we currently use the sum of fatal and recovered. Maybe that causes a bimodal behavior, since we expect fatal and recovered to have different closing period.

**Comment by lisphilar** [comment]
Thank you, we can use "Confirmed minus Faral" and "Recovered" rather than "Confirmed" and "Recovered plus Fatal".

**Comment by Inglezos** [comment]
I don't think that this solution leads to a better result either though. It's the same more or less.

Maybe for calculating the closing period, would be better to find a in-between solution, to the current method with pivot table and the linelist. And the final closing period could be a mean value or something like that.

Perhaps the linelist data are more accurate than the publicly reported recovered cases per country, because they describe specific patients history.
I suspect that since recent months, the recovered cases are reported with some delay period, because of the 14-day quarantine-confirmation period, which could camouflage the real recovering period, between symptoms onset/positive test and recovery. So for example, one individual could be infected on day0, have positive test on day4, develop symptoms on day7 and be better on day 16 (no fever or coughing), with total recovery on day20 (not single symptom besides a little fatigue maybe). Normally this example's recovery period would be 16-4 = 12 days or 20-4 = 16 days. So 12 or 14 is acceptable for using as closing period in our analysis. But sometimes the officially reported recovered case could be several days (i.e. 10 days in average) _after_ the symptoms have disappeared, so that would mean 20+10 = 30 and thus 30-4=26 days, which is very close to the currently calculated modus day.

I don't know for sure if this is actually what is happening, but I suspect that there is a _latency between the actual and reported recovered date_.

**Comment by lisphilar** [comment]
#364 was merged to sort recovered values.

**Comment by lisphilar** [comment]
Note:
With linelist data, recovery/fatal period, not closing period, were as follows with unimodal distribution.

(Median)
Recovery period: 12 days
Fatal period: 4 days

```Python
from matplotlib import pyplot as plt
import covsirphy as cs
data_loader = cs.DataLoader()
linelist = data_loader.linelist()
# Recovery period
r_df = linelist.closed(outcome="Recovered")
r_diff_series = (r_df["Recovered_date"] - r_df["Confirmation_date"]).dt.days
r_diff_series.plot.kde()
plt.show()
r_diff_series.median()
```
https://gist.github.com/lisphilar/e41f58f7f6c9445a8523b6029bf202c6

**Comment by Inglezos** [comment]
Yes I think 12 days is more representative of the recovery period for the mild symptoms cases. What are your thoughts on my detailed comment above from one week ago?
I think there are two recovery periods but I don't know how we can differentiate between these and which one to choose.

**Comment by lisphilar** [comment]
With this analysis, we can add two arguments `recovery_period=12` and `fatal_period=4` to `JHUData._complement_recovered_full()` and revise codes of this method.

**Comment by Inglezos** [comment]
Yes linelist data should be used in closing period but not as hardcoded value, but on-the-fly calculation, since 12 is dynamic.

**Comment by lisphilar** [comment]
Sorry, `fatal_period` is un-necessary for complement.

We can use linelist data for complement, but we need to force users to download linelist data with `DataLoader.jhu()`. This makes users confused and we may need to document it as stdout of `DataLoader.jhu()`.

**Comment by Inglezos** [comment]
Could we do this internally without extra user action? Otherwise yes, we should document this as prerequisite.

**Comment by lisphilar** [comment]
With #384,
-  `LinelistData.recovery_period()` calculates median value of recovery period.
- `DataLoader.jhu()` will download linelist data as well as the dataset of COVID-19 Data Hub are returns `JHUData` instance, calling `DataLoader.linelist()` internally.
- `DataLoader.jhu()` will set `.recovery_period` property with the value of recovery period [days] calculated with `LinelistData.recovery_period()`.
- `JHUData._complement_recovered_full()` uses the expected value of recovery period set with `.recovery_period` property.

**Comment by lisphilar** [comment]
Can we close this issue?

**Comment by Inglezos** [comment]
Yes yes we can close this.

**Comment by Inglezos** [comment]
On second revision, I think it is a good idea to use `self._recovery_period` combined with `calculate_closing_period()` instead for the final recovery period value. That is because valid linelist data are too few (only 277 values), compared to the total cases, they might not be very representative. That coimbination would mean for example an averaging of their mode values or averaging of their mean values or something like that.

**Comment by lisphilar** [comment]
When `JHUData` instance is created by `DataLoader`, we always use recovery period calculated by linelist.
When `JHUData` instance is created by class method `JHUData.from_dataframe()` with local CSV file, we use `.calculate_closing_period()`.

This should be discussed in a new issue, but combination of recovery period calculated by linelist and recovery period (not closing period) by JHU dataset will be a good idea.
